rust/src/detect/transforms/base64.rs

```diff
@@ -223,215 +186,287 @@ fn parse_transform_base64(
-                let s_val = unsafe { CStr::from_ptr(self.offset_str) };
-                let o_val = unsafe { CStr::from_ptr(other.offset_str) };
-                res = s_val == o_val;
-            } else if !self.offset_str.is_null() || !other.offset_str.is_null() {
-                return false;
-            }
+    let ctx = cast_pointer!(ctx, DetectTransformFromBase64Data);
 
-            res && self.nbytes == other.nbytes
-                && self.flags == other.flags
-                && self.offset == other.offset
-                && self.mode == other.mode
+    if ctx.offset > 0 {
+        if ctx.offset >= input_len {
+            return;
+        }
+        input = &input[ctx.offset as usize..];
+    }
+    if ctx.nbytes > 0 {
+        if ctx.nbytes as usize >= input.len() {
+            return;
         }
+        input = &input[..ctx.nbytes as usize];
     }
 
-    fn valid_test(
-        args: &str, nbytes: u32, nbytes_str: &str, offset: u32, offset_str: &str,
-        mode: SCBase64Mode, flags: u8,
-    ) {
-        let tbd = SCDetectTransformFromBase64Data {
-            flags,
-            nbytes,
-            nbytes_str: if !nbytes_str.is_empty() {
-                CString::new(nbytes_str).unwrap().into_raw()
-            } else {
-                std::ptr::null_mut()
-            },
-            offset,
-            offset_str: if !offset_str.is_empty() {
-                CString::new(offset_str).unwrap().into_raw()
-            } else {
-                std::ptr::null_mut()
-            },
-            mode,
-            serialized: std::ptr::null_mut(),
-            serialized_len: 0,
-        };
+    let output_len = get_decoded_buffer_size(input.len() as u32);
+    // no realloc, we only can shrink
+    let output = SCInspectionBufferCheckAndExpand(buffer, output_len);
+    if output.is_null() {
+        // allocation failure
+        return;
+    }
 
-        let (_, val) = parse_transform_base64(args).unwrap();
-        assert_eq!(val, tbd);
+    let num_decoded = SCBase64Decode(input.as_ptr(), input.len(), ctx.mode, output);
+    if num_decoded > 0 {
+        SCInspectionBufferTruncate(buffer, num_decoded);
     }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn DetectTransformFromBase64DecodeRegister() {
+    let kw = SCTransformTableElmt {
+        name: b"from_base64\0".as_ptr() as *const libc::c_char,
+        desc: b"convert the base64 decode of the buffer\0".as_ptr() as *const libc::c_char,
+        url: b"/rules/transforms.html#from_base64\0".as_ptr() as *const libc::c_char,
+        Setup: Some(base64_setup),
+        flags: SIGMATCH_OPTIONAL_OPT,
+        Transform: Some(base64_transform),
+        Free: Some(base64_free),
+        TransformValidate: None,
+        TransformId: Some(base64_id),
+    };
+    unsafe {
+        G_TRANSFORM_BASE64_ID = SCDetectHelperTransformRegister(&kw);
+        if G_TRANSFORM_BASE64_ID < 0 {
+            SCLogWarning!("Failed registering transform base64");
+        }
+    }
```

Comment:
Nested unsafe needed?

Reply:
This was copy pasted from xor transform

Is there a linter that should catch these ?